nengo_loihi/simulator.py

```diff
@@ -487,7 +488,25 @@ def run_steps(self, steps):
             raise SimulatorClosed("Simulator cannot run because it is closed.")
         if self._run_steps is None:
             self._make_run_steps()
-        self._run_steps(steps)
+        try:
+            self._run_steps(steps)
+        except Exception:
+            if "loihi" in self.sims:
+                h2c = self.sims["loihi"].nengo_io_h2c
+                c2h = self.sims["loihi"].nengo_io_c2h
+
+                printed = False
+                for _ in range(steps):
+                    h2c.write(h2c.numElements, [0] * h2c.numElements)
+                    c2h.read(c2h.numElements)
+                    if not printed:
+                        print(traceback.format_exc())
+                        print("\nAttempting to end simulation...")
+                        printed = True
+                self.sims["loihi"].n2board.nxDriver.stopExecution()
+                self.sims["loihi"].n2board.nxDriver.stopDriver()
+                self.sims["loihi"].wait_for_completion()
```

Comment:
Oh, I forgot to say: even though it doesn't make a difference, I think we should call this before we stop excecution and the driver. Intel said that's the way it should go, so I would air in that direction. Also, I think the `n2board` calls should be done by the Loihi simulator, maybe in `disconnect`.

Reply:
I put up a fix for this, and found that I can actually get rid of the hang on exit as well, so that when we get an exception, the simulation exits as we would expect.

Unfortunately, there are still some hangs (for example, with the learning tests), that happen in a different place and aren't solved by this, but they weren't solved before my fixup, either.

Reply:
I should mention, that I think the learning hang is because we don't combine errors properly when we have more than one input to a learning rule. Rather than summing the inputs, we concatenate them so we end up with too many error values that we then try to send to the chip, and I think this causes the hang. So that one might not a Python exception at all, so it's not surprising that nothing in this branch helps with it.